Converted Library from a Unity Project into a Unity Package

paulhayes opened the pull request:
This totally reorganises the project. But this will mean it's much easier to add into other Unity projects. I've followed Unity's guide here:

https://docs.unity3d.com/Manual/CustomPackages.html
https://docs.unity3d.com/Manual/cus-layout.html
https://docs.unity3d.com/Manual/upm-manifestPkg.html

Having it as a package is good for two main reasons:
* The project will no longer be tied to a specific Unity version
* It can be installed very easily via any of the following methods:
  *  by adding repository url in Unity's Package Manager
  * downloading the repo into the Packages directory of a project
  * cloning the repo into the Packages directory of a project ( this is best when you are developing the Package )
  * cloning the repo and adding using the Add Local button in the Package Manager.

I'm hoping your willing to make this pull request. But understand if it's too radical from main, if so could you make a new branch specifically for the package variant, and I'll make a new pull request for that branch. It's possible to reference a specific branch when adding a Unity Package via url.

And thank you @velorexe, for making such a great library. 🥳

Velorexe commented:
Wow! This is a great PR!

I saw your fork yesterday morning and looked at the progress and was hoping to see a PR rolling in soon :p
For now, since the package is primarily used by people using the exported `.unitypackage`, and I see that the code logic stays the same, it can be merged to `main`. I'll have to take a look how I'm going to merge the experimental `develop` branch into it, but that's a problem for me to figure out.

I do have one question, since I couldn't find it in the documentation, will Android resources (`*.jar` files, custom manifests, etc.) be brought into the build?

paulhayes commented:
That's a good question. Let me check  before you do the pull.

We may need a post package import script that asks the user if they want to
add these attributes to the manifest.

I wish the way Unity handled android manifests was less fragile, it's
something that's bothered me in the past.



On Wed, 13 Mar 2024, 08:29 David, ***@***.***> wrote:

> Wow! This is a great PR!
>
> I saw your fork yesterday morning and looked at the progress and was
> hoping to see a PR rolling in soon :p
> For now, since the package is primarily used by people using the exported
> .unitypackage, and I see that the code logic stays the same, it can be
> merged to main. I'll have to take a look how I'm going to merge the
> experimental develop branch into it, but that's a problem for me to
> figure out.
>
> I do have one question, since I couldn't find it in the documentation,
> will Android resources (*.jar files, custom manifests, etc.) be brought
> into the build?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Velorexe/Unity-Android-Bluetooth-Low-Energy/pull/40#issuecomment-1993810737>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAE3KT2YLIJQ7O2LZ6VBYY3YYAE5RAVCNFSM6AAAAABESWQDQKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOJTHAYTANZTG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> com>
>

paulhayes commented:
You're absolutely right of course. For the package to work, the manifest needs to be moved to main Assets/ directory. So I have two suggestions of a solution.

When the package has finished being installed, it opens an EditorWindow and prompt the user to make one of the following choices:
* If the Manifest.xml doesn't exist in Assets, copy the default one across from the package
* If the Manifest.xml already does exist. Append the necessary features to the existing Manifest.xml
* Do nothing and let the user resolve this manually.

We could also check which Android API they are using. If Android API 23 or later, prompt them that they will need to use runtime permissions to correctly scan and connect to a bluetooth device.

If you like that solution. I can add it to my repo and make a new pull request.

Velorexe commented:
The `EditorWindow` is a good idea to give the user their own control on what they can do. I think merging the manifests with the permissions and things that are needed will get quite convoluted, so prompting them with `You need to include permission X and thing Y.` with a copy-able field is the right way to go (like you suggested). If they don't own a custom manifest (have to check if you can get programmatically), then copy the one from the package.

The Android API 23 is an interesting one, not sure if it's the responsibility of the package to prompt these runtime permissions or if it's up to the user themselves to implement it. If it's the responsibility of the package, then it's better to add it as a `boolean` property to the `BleManager`. In the `develop` branch I'm already checking the API version on the device that the app is deployed on, so it'd only be a temporary improvement.

If you can add that to this PR (or a different one) then that would be great! Else I can also chip in with the ideas I mentioned in this comment, you can let me know.

paulhayes commented:
Hi @Velorexe , I've updated my pull request.

I've added a Configuration Wizard, it opens automatically on package import, and it also can be opened from Window->"Android Bluetooth Low Energy Library"

If you don't already have an AndroidManifest.xml installed in Assets, it will prompt you with a button that copies across the template AndroidManifest.

The Wizard will display which permissions you should have and whether they are in the Manifest. Depending on the Projects target android SDK this is either the older Android permissions, or both the legacy and latest permissions for Bluetooth.

There's also a few other quality of life features, like a link to the Android Permission documenation in the wizard.

paulhayes commented:
Did you get a chance yet to review @Velorexe? Happy to make any changes you'd suggest.

paulhayes commented:
By the way @Velorexe , today I realised that my pull request would cause problems for Unity Projects 2019.4 and below. 

I've added some compilation directives so that both my new AppPermission runtime feature, and PostInstallWizard run properly in 2019.4 and below. I'm glad I spotted this before you accepts my pull request.

paulhayes commented:
> I'll pull the changes and check them out later today, thanks for going through the effort to make these wizards!

Hey, how are you getting on with this? I know you must feel swamped as the library has got a few recent bug reports and requests for help. Let me know your thoughts. 

I'm really keen to help out with maintaining and helping support this library. You've written a really solid foundation and I want to help keep it useful to as wide a group as possible.

Velorexe commented:
Hi @paulhayes 

Thanks for the reminder on this! I haven't checked your PR yet since I'm indeed swamped with some other work as well, so I've put maintaining open-source repos to the side for a while. I'll make sure to test and merge it this weekend, as I'm going on a long vacation after this week.

And yeah for sure! If you have contributions then please share them. The alpha build was a side project I was testing out, but having an extra maintainer / contributor is always nice. I would really appreciate it!

Velorexe reviewed (commented):
Most of it looks really good! Bug fixes, great improvements and all; definitely something I want in the package.

Velorexe commented on Editor/PostInstallWizard.cs:
```diff
@@ -0,0 +1,190 @@
+#endif
+
+        public static void RefreshFiles(){
+            assetsAndroidManifestPath = Path.Combine(ProjectPath(),"Plugins/Android/AndroidManifest.xml");
```
Not needed, as custom Android Manifests will always be in the same location.

paulhayes commented on Editor/PostInstallWizard.cs:
```diff
@@ -0,0 +1,190 @@
+                    var rect = EditorGUILayout.GetControlRect(GUILayout.Width(height-4));
+                    rect.height = height-4;
+                    rect.y+=2;
+                    GUI.DrawTexture(rect,hasPerm ? m_tickImage : m_crossImage);
```
This sounds like a much better idea than using a external texture. I'll implement immediately.

paulhayes replied:
done.

However it wasn't as straightforward as you might think, because Unity in their wisdom keep messing around with the availbile icon names in different Unity versions. I've settled on two for the tick. But as I was on 2021.3 the ones you suggested weren't available.

So we will have to keep an eye on this in future Unity releases, as there's no way to know if they are changing the name of an icon. It seems they have in the past.

Velorexe replied:
Good point, I didn't check for cross-version compatibility. Hopefully people will create an issue in case something pops up on newer versions. For now, this looks good and I'll approve / merge it!

paulhayes commented on Editor/PostInstallWizard.cs:
```diff
@@ -0,0 +1,190 @@
+                EditorGUILayout.EndHorizontal();
+            }
+            EditorGUILayout.GetControlRect();
+            EditorGUILayout.TextField("Click here <a href=\"https://developer.android.com/develop/connectivity/bluetooth/bt-permissions\">more information on bluetooth permissions</a>", style);
```
done

paulhayes commented:
Is there anything left to do on this before it can be merged into main?

I think I resolved any remaining issues.

Velorexe commented:
Seen your latest commit, looks good! I'll merge it.
